### PR TITLE
Add content header if the Request doesnot has one for CRT sync Http clients

### PR DESCRIPTION
**Pull request by joviegas**
## Motivation and Context

For request with blob payload
https://github.com/aws/aws-sdk-java-v2/blob/6671c79f763405e245bf96c299061c17e22f49ab/services/sagemakerruntime/src/main/resources/codegen-resources/service-2.json#L300-L302

- The signature mismatch error was seen at the server side since the content-length header was not present

- This issue is only seen for AWS Sync CRT client.
- Other clients like url. Apache added the content length at the end , just before sending the request

## Modifications

- Added code to calculate and add content length for Binary payload and String payload in Marshallers

## Testing

- Added Junits

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/request/CrtRequestAdapter.java [hunk 1]
```diff
@@ -131,11 +134,22 @@ private static List<HttpHeader> createHttpHeaderList(URI uri, HttpExecuteRequest
...
 
-        // Add the rest of the Headers
         sdkRequest.forEachHeader((key, value) -> value.stream().map(val -> new HttpHeader(key, val)).forEach(crtHeaderList::add));
 
         return crtHeaderList;
     }
+
+    public static long calculateContentLength(ContentStreamProvider contentStreamProvider) {
+        try {
+            return  IoUtils.toByteArray(contentStreamProvider.newStream()).length ;
```
Comment (L-Applin) [inline http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/request/CrtRequestAdapter.java:150]: Wont that buffer the whole body content in memory? Do we really want to do that?

Comment (joviegas) [reply http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/request/CrtRequestAdapter.java:150]: Need to rework on this , due to internal comment from Matthew miller to handle this in Marshallers.  Will update the PR with that approach.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java [hunk 2]
```diff
@@ -181,12 +181,16 @@ void doMarshall(SdkPojo pojo) {
             Object val = field.getValueOrDefault(pojo);
             if (isExplicitBinaryPayload(field)) {
                 if (val != null) {
-                    request.contentStreamProvider(((SdkBytes) val)::asInputStream);
+                    SdkBytes sdkBytes = (SdkBytes) val;
+                    request.contentStreamProvider(sdkBytes::asInputStream);
+                    updateContentLengthHeader(sdkBytes.asByteArray().length);
```
Comment (millems) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java:186]: `asByteArrayUnsafe()`?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java [hunk 3]
```diff
@@ -196,6 +200,12 @@ void doMarshall(SdkPojo pojo) {
         }
     }
 
+    private void updateContentLengthHeader(int contentLength) {
+        if (!request.firstMatchingHeader(CONTENT_LENGTH).isPresent()) {
+            request.putHeader(CONTENT_LENGTH, Integer.toString(contentLength));
+        }
```
Comment (millems) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java:206]: When will this condition be needed? Why not just set it?

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/XmlProtocolMarshaller.java [hunk 4]
```diff
@@ -91,8 +91,11 @@ void doMarshall(SdkPojo pojo) {
             Object val = field.getValueOrDefault(pojo);
 
             if (isBinary(field, val)) {
-                request.contentStreamProvider(((SdkBytes) val)::asInputStream);
+                SdkBytes sdkBytes = (SdkBytes) val;
+                request.contentStreamProvider(sdkBytes::asInputStream);
                 setContentTypeHeaderIfNeeded("binary/octet-stream");
+                request.putHeader(CONTENT_LENGTH, Integer.toString(sdkBytes.asByteArray().length));
```
Comment (millems) [inline core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/XmlProtocolMarshaller.java:97]: `asByteArrayUnsafe()`?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java (hunk 3 above)
Comment (joviegas) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/internal/marshall/JsonProtocolMarshaller.java]: Thank you for catching that oversight. I assumed that "MergeCustomHeadersStage" would override the headers before this. However, this stage comes after Marshalling is done. I removed this check.

File: core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/XmlProtocolMarshaller.java (hunk 4 above)
Comment (joviegas) [reply core/protocols/aws-xml-protocol/src/main/java/software/amazon/awssdk/protocols/xml/internal/marshall/XmlProtocolMarshaller.java]: done

File: test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/contentlength/MarshallersAddContentLengthTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+import software.amazon.awssdk.services.protocolrestjson.ProtocolRestJsonClient;
+import software.amazon.awssdk.services.protocolrestjson.model.OperationWithExplicitPayloadStructureResponse;
+import software.amazon.awssdk.services.protocolrestjson.model.SimpleStruct;
+import software.amazon.awssdk.services.protocolrestxml.ProtocolRestXmlClient;
+import software.amazon.awssdk.services.protocolrestxml.model.OperationWithExplicitPayloadStringResponse;
+
+public class MarshallersAddContentLengthTest {
+    public static final String STRING_PAYLOAD = "TEST_STRING_PAYLOAD";
+    @Rule
+    public WireMockRule wireMock = new WireMockRule(0);
```
Comment (L-Applin) [inline test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/contentlength/MarshallersAddContentLengthTest.java:48]: Could we use JUnit 5 with `@WireMockTest` instead?

Comment (joviegas) [reply test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/contentlength/MarshallersAddContentLengthTest.java]: done

**Review by L-Applin — APPROVED** [review]

**Pull request merged**